src/lib.rs

```diff
@@ -14,19 +14,36 @@
 //! a.out(pio::OutDestination::PINS, 1);
 //! a.jmp(pio::JmpCondition::Always, &mut loop_label);
 //!
-//! let program = a.assemble();
+//! let program = a.assemble(None);
+//! ```
+//!
+//! ## Wrapping
+//! ```rust
+//! let mut a = pio::Assembler::new();
+//!
+//! let mut wrap_source = a.label();
+//! let mut wrap_target = a.label();
+//!
+//! // Initialize pin direction only once
+//! a.set(pio::SetDestination::PINDIRS, 1);
+//! a.bind(&mut wrap_target);
+//! a.out(pio::OutDestination::PINS, 1);
+//! a.bind(&mut wrap_source);
+//!
+//! let program = a.assemble(Some((wrap_source, wrap_target)));
 //! ```
 
 #![no_std]
 // PIO instr grouping is 3/5/3/5
 #![allow(clippy::unusual_byte_groupings)]
 #![allow(clippy::upper_case_acronyms)]
 
-// pub because the proc macro needs it
-#[doc(hidden)]
-pub extern crate alloc;
+use arrayvec::ArrayVec;
 
-use alloc::vec::Vec;
+/// Maximum program size in bytes.
+///
+/// See Chapter 3, Figure 38 for reference of the value.
+const MAX_PROGRAM_SIZE: usize = 32;
```

Comment:
The assembler should not care about this, there could be another pio module released tomorrow with double the state machine memory. If this is just for the ArrayVec, how about passing that into the assembler?

Reply:
I agree. Would the following work for you?
```rust
pub struct Assembler<const N: usize> { ... }
```
This of course still imposes hard-coded limit on the program size, but without it the assembler would need to do some allocations to store its internal state making the usage of this crate more difficult on the actual hardware.

Reply:
I was thinking more like `pub struct Assembler<V, 'v> { instructions: &'v mut V }` and `Assembler::new(v instance)`

Reply:
Right, I see. For that we need to define `VecLike` trait. I hope that implementing that on `Vec` doesn't require `alloc` to be included in the binary as that is what I'm trying to avoid with `ArrayVec`.

Reply:
ah that's a good point. maybe `const N: usize` is the way to go then 😢

Reply:
I think we can actually feature gate the implementation so having this be generic over `Vec` and `ArrayVec` should be possible.

Reply:
I tried to implement this to be generic over `Vec` and `ArrayVec`, but the usage deteriorated greatly as the user would now need to tell everywhere which container to use and with which parameter. Once GATs land, I hope that this could be refactored so that one could at least use `Assembler<VecFamily>` instead of `Assembler<Vec<Instruction>>`. I hope that this is enough for now, and a future PR can refactor this to be more versatile.